ext/dom/parentnode.c

```diff
@@ -0,0 +1,411 @@
+		first = nodep->children;
+
+		while (first && first->type != XML_ELEMENT_NODE) {
+			first = first->next;
+		}
+	}
+
+	if (!first) {
+		ZVAL_NULL(retval);
+		return SUCCESS;
+	}
+
+	php_dom_create_object(first, retval, obj);
+	return SUCCESS;
+}
+/* }}} */
+
+/* {{{ lastElementChild DomParentNode
+readonly=yes
+URL: https://www.w3.org/TR/dom/#dom-parentnode-lastelementchild
+*/
+int dom_parent_node_last_element_child_read(dom_object *obj, zval *retval)
+{
+	xmlNode *nodep, *last = NULL;
+
+	nodep = dom_object_get_node(obj);
+
+	if (nodep == NULL) {
+		php_dom_throw_error(INVALID_STATE_ERR, 0);
+		return FAILURE;
+	}
+
+	if (dom_node_children_valid(nodep) == SUCCESS) {
+		last = nodep->last;
+
+		while (last && last->type != XML_ELEMENT_NODE) {
+			last = last->prev;
+		}
+	}
+
+	if (!last) {
+		ZVAL_NULL(retval);
+		return SUCCESS;
+	}
+
+	php_dom_create_object(last, retval, obj);
+	return SUCCESS;
+}
+/* }}} */
+
+/* {{{ childElementCount DomParentNode
+readonly=yes
+https://www.w3.org/TR/dom/#dom-parentnode-childelementcount
+*/
+int dom_parent_node_child_element_count(dom_object *obj, zval *retval)
+{
+	xmlNode *nodep, *first = NULL;
+	zend_long count = 0;
+
+	nodep = dom_object_get_node(obj);
+
+	if (nodep == NULL) {
+		php_dom_throw_error(INVALID_STATE_ERR, 0);
+		return FAILURE;
+	}
+
+	if (dom_node_children_valid(nodep) == SUCCESS) {
+		first = nodep->children;
+
+		while (first != NULL) {
+			if (first->type == XML_ELEMENT_NODE) {
+				count++;
+			}
+
+			first = first->next;
+		}
+	}
+
+	ZVAL_LONG(retval, count);
+
+	return SUCCESS;
+}
+/* }}} */
+
+xmlNode* dom_zvals_to_fragment(php_libxml_ref_obj *document, xmlNode *contextNode, zval *nodes, int nodesc)
+{
+	int i;
+	xmlDoc *documentNode;
+	xmlNode *fragment;
+	xmlNode *newNode;
+	zend_class_entry *ce;
+	dom_object *newNodeObj;
+	int stricterror;
+
+	if (contextNode->type == XML_DOCUMENT_NODE || contextNode->type == XML_HTML_DOCUMENT_NODE) {
+		documentNode = (xmlDoc *) contextNode;
+	} else {
+		documentNode = contextNode->doc;
+	}
+
+	fragment = xmlNewDocFragment(documentNode);
+
+	if (!fragment) {
+		return NULL;
+	}
+
+	stricterror = dom_get_strict_error(document);
+
+	for (i = 0; i < nodesc; i++) {
+		if (Z_TYPE(nodes[i]) == IS_OBJECT) {
+			ce = Z_OBJCE(nodes[i]);
+
+			if (instanceof_function(ce, dom_node_class_entry)) {
+				newNodeObj = Z_DOMOBJ_P(&nodes[i]);
+				newNode = dom_object_get_node(newNodeObj);
+
+				if (newNode->doc != documentNode) {
+					php_dom_throw_error(WRONG_DOCUMENT_ERR, stricterror);
+					return NULL;
+				}
+
+				if (newNode->parent != NULL) {
+					xmlUnlinkNode(newNode);
+				}
+
+				newNodeObj->document = document;
+				xmlSetTreeDoc(newNode, documentNode);
+
+				if (!xmlAddChild(fragment, newNode)) {
+					php_dom_throw_error(HIERARCHY_REQUEST_ERR, stricterror);
+					return NULL;
+				}
+
+				continue;
+			} else {
+				zend_type_error("Invalid argument type must be either DOMNode or string");
```

Comment:
Should return here (probably after destroying the fragment). Also for the exception below.